discovery_engine_core/core/src/stack/ops/common.rs

```diff
@@ -12,58 +12,32 @@
 // You should have received a copy of the GNU Affero General Public License
 // along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
-use futures::{stream::FuturesUnordered, StreamExt};
 use tokio::task::JoinHandle;
-use xayn_discovery_engine_providers::Market;
 
 use crate::engine::GenericError;
 
 type ItemsResult<I> = Result<Vec<I>, GenericError>;
-type Requests<I> = FuturesUnordered<JoinHandle<ItemsResult<I>>>;
-
-async fn request_new_items<I: Send>(
-    requests_fn: impl FnOnce() -> Requests<I> + Send,
-) -> ItemsResult<I> {
-    let mut requests = requests_fn();
-    let mut items = Vec::new();
-    let mut error = None;
-
-    while let Some(handle) = requests.next().await {
-        // should we also push handle errors?
-        if let Ok(result) = handle {
-            match result {
-                Ok(batch) => items.extend(batch),
-                Err(err) => {
-                    error.replace(err);
-                }
-            }
-        }
-    }
-
-    if items.is_empty() && error.is_some() {
-        Err(error.unwrap(/* nonempty error */))
-    } else {
-        Ok(items)
-    }
-}
+type Request<I> = JoinHandle<ItemsResult<I>>;
 
 pub(super) async fn request_min_new_items<I: Send>(
     max_requests: u32,
     min_articles: usize,
-    requests_fn: impl Fn(u32) -> Requests<I> + Send + Sync,
+    request_fn: impl Fn(u32) -> Request<I> + Send + Sync,
     filter_fn: impl Fn(Vec<I>) -> ItemsResult<I> + Send + Sync,
 ) -> ItemsResult<I> {
-    let mut items = Vec::new();
+    let mut items = Vec::with_capacity(min_articles);
     let mut error = None;
 
     for request_num in 0..max_requests {
-        match request_new_items(|| requests_fn(request_num)).await {
+        match request_fn(request_num).await {
             // if the API doesn't return any new items, we stop requesting more pages
-            Ok(batch) if batch.is_empty() => break,
-            Ok(batch) => items.extend(batch),
-            Err(err) => {
+            Ok(Ok(batch)) if batch.is_empty() => break,
```

Comment:
🔧 we can pass the page size to this function and check if `batch < page_size` and break earlier to avoid a request to the next empty page. that was a problem before that we for example got 10 items on page 1 and then we did another request because `batch` was not empty.